### PR TITLE
feat: stabilize 0.2.0

### PR DESCRIPTION
**Pull request by Freyskeyd**
# Description

This PR aims at stabilizing the next `0.2.0` during the `kurtosis` test phase.

## Fixes

- Fixes related to the `agglayer-prover` that needed a `native-gnark` setup

## Additions and Changes

- Move part of the `orchestrator` logic into a small `network` task, this is a first step to make the orchestrator lightweight as discuss with @iljakuklic  
- Adding two kind of errors for the `Certifier`
- Rework the epoch calculation in the `ClockRef`
 
### New feature (non-breaking change which adds functionality)

- A new RPC to expose the latest known certificate for a network
- An attempt at implementing `GpuProver`, this is currently not supported as it requires more work.

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

File: crates/agglayer-storage/src/columns/latest_settled_certificate_per_network/mod.rs [hunk 1]
```diff
@@ -17,8 +17,13 @@ mod tests;
 /// | `NetworkId` | (`CertificateId`, `Height`, `EpochNumber`) |
```
Comment (hadjiszs) [inline crates/agglayer-storage/src/columns/latest_settled_certificate_per_network/mod.rs:17]:
```suggestion
/// | `NetworkId` | (`CertificateId`, `Height`, `EpochNumber`, `CertificateIndex`) |
```

File: crates/agglayer-certificate-orchestrator/src/network_task.rs [hunk 2]
```diff
@@ -0,0 +1,995 @@
...
+            // If the certificate is already proven or candidate, it means that the
+            // certification process has already been initiated but not completed.
+            // It also means that the proof exists and thus we should redo the native
+            // execution to update the local state.
+            CertificateStatus::Proven | CertificateStatus::Candidate => {
+                // Redo native execution to get the new_state
+
+                error!(
+                    hash = certificate_id.to_string(),
+                    "CRITICAL: Certificate {certificate_id} is already proven or candidate while \
```
Comment (hadjiszs) [inline crates/agglayer-certificate-orchestrator/src/network_task.rs:197]:
```suggestion
                    "CRITICAL: Certificate {certificate_id} is already proven or candidate \
```

File: crates/agglayer-certificate-orchestrator/src/lib.rs [hunk 3]
```diff
@@ -346,502 +411,72 @@ impl<C, E, CertifierClient, PendingStore, EpochsStore, PerEpochStore, StateStore
...
-            Some(cursor_height) => {
-                warn!(
-                    "Received a certificate with an unexpected height: {} for network {} which is \
-                     currently at {}",
-                    height, network, cursor_height
-                );
-
-                return Err(());
-            }
+            Err(_) => todo!(),
```
Comment (hadjiszs) [inline crates/agglayer-certificate-orchestrator/src/lib.rs:450]: remaining panic, maybe replacing with at least a error log for now

**Review by hadjiszs — APPROVED** [review]

File: crates/agglayer-clock/src/time.rs [hunk 4]
```diff
@@ -76,7 +76,7 @@ impl TimeClock {
 
         // Compute the current Block height and Epoch number
         let current_block = self.update_block_height();
-        let current_epoch = self.calculate_epoch_number(current_block);
+        let current_epoch = Self::calculate_epoch_number(current_block, *self.epoch_duration);
```
Comment (iljakuklic) [inline crates/agglayer-clock/src/time.rs:79]: I am a bit confused here. The epoch is calculated from block number even for time-based clock?

**Review by iljakuklic — APPROVED** [review]

File: crates/agglayer-certificate-orchestrator/src/lib.rs [hunk 5]
```diff
@@ -217,41 +250,69 @@ impl<C, E, CertifierClient, PendingStore, EpochsStore, PerEpochStore, StateStore
...
+                sender.clone()
+            } else {
+                warn!("Unable to find the network task for network {}", network_id);
+                continue;
+            };
+
+            let sender = sender.clone();
+            tokio::spawn(async move {
+                if sender
+                    .send(NewCertificate {
```
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/agglayer-certificate-orchestrator/src/lib.rs:303]: yes I don't want this function to be async as it is called directly in the Poll

File: crates/agglayer-clock/src/time.rs (hunk 4 above)
Comment (Freyskeyd) [reply crates/agglayer-clock/src/time.rs:79]: The `ClockRef` is always calculating the epoch based on the agglayer blocks.
In the time-based clock, a "block" is basically one second.

File: crates/agglayer-certificate-orchestrator/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (iljakuklic) [reply crates/agglayer-certificate-orchestrator/src/lib.rs:303]: > yes I don't want this function to be async as it is called directly in the Poll

In that case, I'd recommend to just use an unbounded channel. It does not need `.await` on send. In effect, this emulates an unbounded channel by spawning a potentially unbounded number of tasks that get suspended on send if the channel is full.

File: crates/agglayer-certificate-orchestrator/src/network_task.rs [hunk 6]
```diff
@@ -0,0 +1,995 @@
...
+            local_state: LocalNetworkStateData::default(),
+            certification_notifier,
+            clock_ref,
+            pending_state: None,
+            certificate_stream,
+            at_capacity_for_epoch: false,
+        })
+    }
+
+    pub(crate) async fn run(mut self) -> Result<NetworkId, Error> {
```
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/agglayer-certificate-orchestrator/src/network_task.rs:88]: Yeah, the idea was to make the task able to be killed and restarted, the `NetworkId` is used to remove the task from the tracking map on the orchestrator. Currently, I can remove this as you suggested.

File: crates/agglayer-certificate-orchestrator/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/agglayer-certificate-orchestrator/src/lib.rs:303]: Having an unbounded channel obfuscate if there is a problem with the network task. We could stack a lot of certificate without noticing it. I don't like the idea of having something that can OOM the node like that, I will switch to a sync version using permit etc

[the remark this answers is not in the document]
Comment (iljakuklic) [reply crates/agglayer-certificate-orchestrator/src/lib.rs:303]: > Having an unbounded channel obfuscate if there is a problem with the network task. We could stack a lot of certificate without noticing it. I don't like the idea of having something that can OOM the node like that, I will switch to a sync version using permit etc

Fair. May not have been clear about this but my point was merely it's preferable to use an unbounded channel to spawning unbounded tasks as a workaround to the `async` infection issue. If you have a proper solution, even better.

**Pull request merged**